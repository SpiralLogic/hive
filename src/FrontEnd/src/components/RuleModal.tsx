import { FunctionComponent, h } from 'preact';
import { useReducer } from 'preact/hooks';
import AntRules from './rules/AntRules';
import BeetleRules from './rules/BeetleRules';
import GrasshopperRules from './rules/GrasshopperRules';
import Modal from './Modal';
import QueenRules from './rules/QueenRules';
import SpiderRules from './rules/SpiderRules';

type Props = { setShowRules: (value: boolean) => void };

const RuleModal: FunctionComponent<Props> = (props) => {
  const ruleList = [<QueenRules />, <BeetleRules />, <SpiderRules />, <GrasshopperRules />, <AntRules />];
  const changeRule = (currentRuleIndex: number, { type }: { type: 'next' | 'prev' }): number => {
    if (type === 'next') return ++currentRuleIndex % ruleList.length;
    if (type === 'prev' && currentRuleIndex > 0) return --currentRuleIndex % ruleList.length;
    return ruleList.length - 1;
  };
  const [currentRuleIndex, changeCurrentRule] = useReducer(changeRule, 0);
  return (
    <Modal name="rules" onClose={() => props.setShowRules(false)}>
      <div class="menu">
        <button title="Previous" onClick={() => changeCurrentRule({ type: 'next' })}>
          &lt;
        </button>
        <button title="Next" onClick={() => changeCurrentRule({ type: 'prev' })}>
          &gt;
        </button>
      </div>
      {ruleList[currentRuleIndex]}
    </Modal>
  );
};

RuleModal.displayName = 'Rules';
export default RuleModal;
