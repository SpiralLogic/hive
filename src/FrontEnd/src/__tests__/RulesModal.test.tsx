import { h } from 'preact';
import { renderElement } from './helpers';
import RuleModal from '../components/RuleModal';

describe('Rules snapshot tests', () => {
  test('snapshot', () => {
    expect(renderElement(<RuleModal setShowRules={() => ({})} />)).toMatchSnapshot();
  });
});
