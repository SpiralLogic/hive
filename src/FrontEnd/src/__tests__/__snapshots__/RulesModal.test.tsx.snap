// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rules tests snapshot 1`] = `
<div
  class="modal"
>
  <div
    class="rules"
    role="dialog"
    title="rules"
  >
    <div
      class="menu"
    >
      <button
        title="Previous"
      >
        &lt;
      </button>
      <button
        title="Next"
      >
        &gt;
      </button>
    </div>
    <div
      class="left"
      role="grid"
    >
      <div
        role="row"
      >
        <div
          class="correct"
          role="cell"
        >
          <svg
            viewBox="0 0 100 100"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              href="#hex"
            />
          </svg>
          <div
            class="tile"
          >
            <svg
              viewBox="0 0 100 100"
              xmlns="http://www.w3.org/2000/svg"
            >
              <use
                href="#hex"
              />
            </svg>
            <span>
              <span>
                ✓
              </span>
            </span>
          </div>
        </div>
        <div
          role="cell"
        >
          <svg
            viewBox="0 0 100 100"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              href="#hex"
            />
          </svg>
          <div
            class="tile creature"
          >
            <svg
              viewBox="0 0 100 100"
              xmlns="http://www.w3.org/2000/svg"
            >
              <use
                href="#hex"
              />
              <use
                class="creature ant"
                href="#ant"
              />
            </svg>
            
          </div>
        </div>
        <div
          role="cell"
        >
          <svg
            viewBox="0 0 100 100"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              href="#hex"
            />
          </svg>
          <div
            class="tile creature"
          >
            <svg
              viewBox="0 0 100 100"
              xmlns="http://www.w3.org/2000/svg"
            >
              <use
                href="#hex"
              />
              <use
                class="creature spider"
                href="#spider"
              />
            </svg>
            
          </div>
        </div>
        <div
          role="none"
        >
          <svg
            viewBox="0 0 100 100"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              href="#hex"
            />
          </svg>
        </div>
      </div>
      <div
        class="start"
        role="row"
      >
        <div
          class="correct"
          role="cell"
        >
          <svg
            viewBox="0 0 100 100"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              href="#hex"
            />
          </svg>
          <div
            class="tile"
          >
            <svg
              viewBox="0 0 100 100"
              xmlns="http://www.w3.org/2000/svg"
            >
              <use
                href="#hex"
              />
            </svg>
            <span>
              <span>
                ✓
              </span>
            </span>
          </div>
        </div>
        <div
          role="cell"
          style="z-index: 1;"
        >
          <svg
            viewBox="0 0 100 100"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              href="#hex"
            />
          </svg>
          <div
            class="selected tile creature"
          >
            <svg
              viewBox="0 0 100 100"
              xmlns="http://www.w3.org/2000/svg"
            >
              <use
                href="#hex"
              />
              <use
                class="creature queen"
                href="#queen"
              />
            </svg>
            
          </div>
          <svg
            class="arrow topLeft correct"
            style="--span: 1;"
            viewBox="0 0 100 100"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              href="#arrow"
            />
          </svg>
          <svg
            class="arrow left correct"
            style="--span: 1;"
            viewBox="0 0 100 100"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              href="#arrow"
            />
          </svg>
          <svg
            class="arrow right correct"
            style="--span: 1;"
            viewBox="0 0 100 100"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              href="#arrow"
            />
          </svg>
          <svg
            class="arrow bottomRight correct"
            style="--span: 1;"
            viewBox="0 0 100 100"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              href="#arrow"
            />
          </svg>
        </div>
        <div
          class="correct"
          role="cell"
        >
          <svg
            viewBox="0 0 100 100"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              href="#hex"
            />
          </svg>
          <div
            class="tile"
          >
            <svg
              viewBox="0 0 100 100"
              xmlns="http://www.w3.org/2000/svg"
            >
              <use
                href="#hex"
              />
            </svg>
            <span>
              <span>
                ✓
              </span>
            </span>
          </div>
        </div>
        <div
          role="cell"
        >
          <svg
            viewBox="0 0 100 100"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              href="#hex"
            />
          </svg>
          <div
            class="tile creature"
          >
            <svg
              viewBox="0 0 100 100"
              xmlns="http://www.w3.org/2000/svg"
            >
              <use
                href="#hex"
              />
              <use
                class="creature spider"
                href="#spider"
              />
            </svg>
            
          </div>
        </div>
      </div>
      <div
        role="row"
      >
        <div
          role="cell"
        >
          <svg
            viewBox="0 0 100 100"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              href="#hex"
            />
          </svg>
          <div
            class="tile creature"
          >
            <svg
              viewBox="0 0 100 100"
              xmlns="http://www.w3.org/2000/svg"
            >
              <use
                href="#hex"
              />
              <use
                class="creature grasshopper"
                href="#grasshopper"
              />
            </svg>
            
          </div>
        </div>
        <div
          class="correct"
          role="cell"
        >
          <svg
            viewBox="0 0 100 100"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              href="#hex"
            />
          </svg>
          <div
            class="tile"
          >
            <svg
              viewBox="0 0 100 100"
              xmlns="http://www.w3.org/2000/svg"
            >
              <use
                href="#hex"
              />
            </svg>
            <span>
              <span>
                ✓
              </span>
            </span>
          </div>
        </div>
        <div
          role="cell"
        >
          <svg
            viewBox="0 0 100 100"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              href="#hex"
            />
          </svg>
          <div
            class="tile creature"
          >
            <svg
              viewBox="0 0 100 100"
              xmlns="http://www.w3.org/2000/svg"
            >
              <use
                href="#hex"
              />
              <use
                class="creature queen"
                href="#queen"
              />
            </svg>
            
          </div>
        </div>
        <div
          role="cell"
        >
          <svg
            viewBox="0 0 100 100"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              href="#hex"
            />
          </svg>
          <div
            class="tile creature"
          >
            <svg
              viewBox="0 0 100 100"
              xmlns="http://www.w3.org/2000/svg"
            >
              <use
                href="#hex"
              />
              <use
                class="creature beetle"
                href="#beetle"
              />
            </svg>
            
          </div>
        </div>
      </div>
      <div
        role="row"
      >
        <div
          role="none"
        >
          <svg
            viewBox="0 0 100 100"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              href="#hex"
            />
          </svg>
        </div>
        <div
          role="cell"
        >
          <svg
            viewBox="0 0 100 100"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              href="#hex"
            />
          </svg>
          <div
            class="tile creature"
          >
            <svg
              viewBox="0 0 100 100"
              xmlns="http://www.w3.org/2000/svg"
            >
              <use
                href="#hex"
              />
              <use
                class="creature grasshopper"
                href="#grasshopper"
              />
            </svg>
            
          </div>
        </div>
        <div
          role="cell"
        >
          <svg
            viewBox="0 0 100 100"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              href="#hex"
            />
          </svg>
          <div
            class="tile creature"
          >
            <svg
              viewBox="0 0 100 100"
              xmlns="http://www.w3.org/2000/svg"
            >
              <use
                href="#hex"
              />
              <use
                class="creature ant"
                href="#ant"
              />
            </svg>
            
          </div>
        </div>
        <div
          role="cell"
        >
          <svg
            viewBox="0 0 100 100"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              href="#hex"
            />
          </svg>
          <div
            class="tile creature"
          >
            <svg
              viewBox="0 0 100 100"
              xmlns="http://www.w3.org/2000/svg"
            >
              <use
                href="#hex"
              />
              <use
                class="creature beetle"
                href="#beetle"
              />
            </svg>
            
          </div>
        </div>
      </div>
    </div>
    <button
      title="Close"
    >
      Close
    </button>
  </div>
</div>
`;
